### ISSUE TITLE
Add rotate function to image-signal

### ISSUE THREAD

**Issue by magnunor** [hyperspy/hyperspy#729]
Currently one can rotate images using the s.map(scipy.ndimage.rotate, ...). However it would be nice having a s.rotate method, since it is a fairly common function people want to do on their TEM/STEM images.

It could simply be a function calling s.map with the required ndimage.rotate parameters, one could also make an interactive version.

**Comment by vidartf** [comment]
I already have implemented interactive rotation for hyperspyUI, and the relevant code is here:
https://github.com/vidartf/hyperspyUI/blob/master/hyperspyui/plugins/imagerotation.py#L58-L164
Moving the actual rotation code to hyperspy makes sense to me. Not sure how easy/difficult it would be to get it interactive directly in hyperpsy.

Major points worth noting:
* Hyperspy gets to skip all the code for inferring rotation axes.
* I avoid using scipy.ndimage.rotate when the rotation is a multiple of 90, to avoid interpolation. This code is not completely mature, so would need to be polished up. If we don't care about that, the relevant code is [here](https://github.com/vidartf/hyperspyUI/blob/master/hyperspyui/plugins/imagerotation.py#L130-L164).
* This code assumes same scale/units for involved axes.
* When rotating 90 degrees, should hyperspy axes "follow" rotation, or stay as before?

**Comment by francisco-dlp** [comment]
I agree that a rotate method will be most welcome. Implementing this should be easy once #606 is ready.

Interactivity: for that we should add an `out` method to `Signal.map`.

I think that, for consistency, the axes should stay fixed also when rotating 90 degrees. There is already a way to "rotate" axes and image 90º: im.as_image((-1,-2)).

**Comment by magnunor** [comment]
One thing to keep in mind when using scipy.ndimage.rotate with s.map(): if reshape=True (which is the default), the signal will be larger compared to the original image, but the s.axes_manager[...].scale will not be changed. Thus leading to the new image being incorrectly scaled.

This is not a problem when using reshape=False.

**Comment by vidartf** [comment]
I'm not sure if I understand why `map` would be needed: `scipy.ndimage.rotate` handles ND data already, which is why it has an `axes` argument.

**Comment by francisco-dlp** [comment]
Sure, it is only needed when `reshape=True` as pointed out by @magnunor. Does HyperSpyUI support `reshape=True`?

**Comment by vidartf** [comment]
Reshape is supported, and is handled like this (modified slightly):
```python
old_shape = sig.data.shape
# ... DO ROTATION HERE
if reshape:
    # get_dimensions_from_data autoreplots, disable temporarily
    old_autoreplot = sig.auto_replot
    sig.auto_replot = False
    sig.get_dimensions_from_data()    # Update axes sizes
    sig.auto_replot = old_autoreplot

    # For axes that has changed size, adjust offsets
    # We assume rotation about the center
    diff = np.array(sig.data.shape) - old_shape
    for i, ax in enumerate(sig.axes_manager._axes):
        dx = diff[i] * 0.5 * ax.scale
        ax.offset -= dx
```
Exactly how axes size is updated is not important (`get_dimensions_from_data` might be overkill), but is pretty straight forward with the shape difference.

**Comment by vidartf** [comment]
An alternative version:
```python
old_shape = sig.data.shape
# ... DO ROTATION HERE
if reshape:
    # For axes that has changed size, adjust size and offsets
    # We assume rotation about the center
    diff = np.array(sig.data.shape) - old_shape
    for i, ax in enumerate(sig.axes_manager._axes):
        dx = diff[i] * 0.5 * ax.scale
        ax.offset -= dx
        ax.size += diff[i]
```

**Comment by francisco-dlp** [comment]
Neat!

**Comment by dnjohnstone** [comment]
What was the ultimate conclusion here? Personally I'm quite happy using map and external rotate methods, but I can see potential advantages to incorporating it into the main code.

One thing I would say is if we do rotation - we may as well do general affine transforms?

**Comment by francisco-dlp** [comment]
+1 on a rotate method including @vidartf code above.

**Comment by thomasaarholt** [comment]
In case anyone is interested, I’ve written a function to rotate a signal by a certain angle, and then crop it to it’s maximum non-black-edges rectangular area. Works for square and rectangular shaped spectrum images.
Take a look (if the link breaks, I'll be uploading it to the examples list anyway)
https://nbviewer.jupyter.org/urls/dl.dropbox.com/s/mtvymqmxdt5h0q7/Rotate%20spectra.ipynb?dl=0
This is the crop method I used: http://stackoverflow.com/a/16778797/1018861

**Comment by magnunor** [comment]
@thomasaarholt, nice! I think that functionality would be very useful to have in a `s.rotate` function. Would remove some of the hassle when rotating images for making figures :)
